refactor(compiler-cli): Add experimental local compilation mode

pmvald opened the pull request:
In this mode the compiler generates code based on each individual source file without using its dependencies. This mode is suitable only for fast edit/refresh during development.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
No change to current behaviour.

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald commented:
Don't know why sub_module samples also got included. Thought there is a way to exclude it, but no way! So I created another [PR](https://github.com/angular/angular/pull/49848) where I just add it to gitignore. See which one makes more sense and close the other one.

JeanMeche commented:
You should close this one yourself then :)

devversion commented on packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:
```diff
@@ -383,10 +383,13 @@ export interface TargetOptions {
    * The default value is 'full'.
    */
-  compilationMode?: 'full'|'partial';
+  compilationMode?: 'full'|'partial'|'experimental-local';
```
Have we considered naming this "experimental-single-file"? Local is quite a broad term, but definitely works, just an idea.

pmvald replied:
Yeah, but there would be a risk that for some reason we cannot achieve single file compilation in Angular, and have to content ourselves with "single compile unit" instead. So, I chose local compilation as a more conservative choice. 

But at the end if we were successful we can change the value to "signle-file" (instead of "local"). This should be a one line change in the helper `isLocalCompilation`

devversion replied:
Sounds good. Like you said, we can always change this in the future, and you also explained "local" in the JSDoc

devversion commented on packages/compiler-cli/src/ngtsc/util/src/local_compilation.ts:
```diff
@@ -0,0 +1,14 @@
+
+import {NgCompilerOptions} from '../../core/api';
+
+/** Cannonical logic as to whether the compilation is in local mode. */
```
typo for canonical, and I'd propose something like:

```suggestion
/** Gets whether local compilation is enabled in the compiler options. */
```

devversion commented on packages/compiler-cli/src/ngtsc/util/BUILD.bazel:
```diff
@@ -9,6 +9,7 @@ ts_library(
     ]),
     deps = [
         "//packages:types",
+        "//packages/compiler-cli/src/ngtsc/core:api",
```
Nit might want to move `isLocalCompilation` to `core/src/` so that `utils` does not have a dependency on core (even if just API)

alxhub reviewed (changes requested):
There are also two commits with identical messages in this PR, which should be squashed into one.

pmvald committed:
refactor(compiler-cli): Add experimental local compilation mode.

In this mode the compiler generates code based on each individual source file without using its dependencies. This mode is suitable only for fast edit/refresh during development.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk reviewed (approved):
reviewed-for: public-api

jessicajaniuk commented:
Caretaker: win_tests are known flakes. This is safe to merge.

dylhunn commented:
This PR was merged into the repository by commit 345dd6d81acc6be15cb1b822afa39d6a8548e0b5.
